### ISSUE TITLE
Unable to build from sources - 0.72.x

### ISSUE THREAD

**Issue by leonchabbey** [react-native-tvos/react-native-tvos#574]
### Description

After making changes to the ReactAndroid code base, I'd like to test it my own project.
During build, the following errors occured:
```
> Task :react-native:packages:react-native:ReactAndroid:hermes-engine:buildCMakeRelease[arm64-v8a][libhermes]
C/C++: ninja: Entering directory `/TestApp/node_modules/react-native/ReactAndroid/hermes-engine/.cxx/Release/491u2i1o/arm64-v8a'
C/C++: /TestApp/node_modules/react-native/sdks/hermes/API/hermes/hermes.cpp:2281:15: error: no matching constructor for initialization of 'jsi::JSError'
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1460:3: note: candidate constructor not viable: no known conversion from 'jsi::Value' to 'facebook::jsi::Runtime &' for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1465:3: note: candidate constructor not viable: no known conversion from 'jsi::Value' to 'std::string' (aka 'basic_string<char>') for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1447:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1451:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1455:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1467:3: note: candidate constructor not viable: requires 1 argument, but 3 were provided
C/C++: /TestApp/node_modules/react-native/sdks/hermes/API/hermes/hermes.cpp:2303:9: error: no matching constructor for initialization of 'jsi::JSError'
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1460:3: note: candidate constructor not viable: no known conversion from 'typename remove_reference<Value &>::type' (aka 'facebook::jsi::Value') to 'facebook::jsi::Runtime &' for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1465:3: note: candidate constructor not viable: no known conversion from 'typename remove_reference<Value &>::type' (aka 'facebook::jsi::Value') to 'std::string' (aka 'basic_string<char>') for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1447:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1451:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1455:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/../jsi/jsi.h:1467:3: note: candidate constructor not viable: requires 1 argument, but 3 were provided
C/C++: /TestApp/node_modules/react-native/sdks/hermes/API/hermes/hermes.cpp:2281:15: error: no matching constructor for initialization of 'jsi::JSError'
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1460:3: note: candidate constructor not viable: no known conversion from 'jsi::Value' to 'facebook::jsi::Runtime &' for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1465:3: note: candidate constructor not viable: no known conversion from 'jsi::Value' to 'std::string' (aka 'basic_string<char>') for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1447:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1451:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1455:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1467:3: note: candidate constructor not viable: requires 1 argument, but 3 were provided
C/C++: /TestApp/node_modules/react-native/sdks/hermes/API/hermes/hermes.cpp:2303:9: error: no matching constructor for initialization of 'jsi::JSError'
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1460:3: note: candidate constructor not viable: no known conversion from 'typename remove_reference<Value &>::type' (aka 'facebook::jsi::Value') to 'facebook::jsi::Runtime &' for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1465:3: note: candidate constructor not viable: no known conversion from 'typename remove_reference<Value &>::type' (aka 'facebook::jsi::Value') to 'std::string' (aka 'basic_string<char>') for 1st argument
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1447:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1451:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1455:3: note: candidate constructor not viable: requires 2 arguments, but 3 were provided
C/C++: /TestApp/node_modules/react-native/ReactCommon/jsi/jsi/jsi.h:1467:3: note: candidate constructor not viable: requires 1 argument, but 3 were provided

> Task :react-native:packages:react-native:ReactAndroid:hermes-engine:buildCMakeRelease[arm64-v8a][libhermes] FAILED
```

Those same errors happen for every target architectures.

### ⚠️  Other tests
-  Building the rn-tester app worked ✅ 
-  Building from sources for a "normal" react-native app (the official repo) worked. ✅ 

### React Native Version

0.72.4-0

### Output of `npx react-native info`

System:
  OS: macOS 13.4.1
  CPU: (8) arm64 Apple M1
  Memory: 154.86 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 16.18.1
    path: ~/.nvm/versions/node/v16.18.1/bin/node
  Yarn:
    version: 1.22.19
    path: ~/.nvm/versions/node/v16.18.1/bin/yarn
  npm:
    version: 8.19.2
    path: ~/.nvm/versions/node/v16.18.1/bin/npm
  Watchman:
    version: 2022.12.26.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.11.3
    path: ~/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 22.4
      - iOS 16.4
      - macOS 13.3
      - tvOS 16.4
      - watchOS 9.4
  Android SDK:
    Android NDK: 21.4.7075529
IDEs:
  Android Studio: 2022.3 AI-223.8836.35.2231.10671973
  Xcode:
    version: 14.3/14E222b
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 11.0.20.1
    path: /Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home/bin/javac
  Ruby:
    version: 2.6.10
    path: ~/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native: Not Found
  react-native-macos: Not Found
  react-native-tvos:
    installed: 0.72.4-0
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: Not found

### Steps to reproduce

Use the following project: [TestApp.zip](https://github.com/react-native-tvos/react-native-tvos/files/12619984/TestApp.zip)
It has been created with `npx react-native@latest init TestApp --template=react-native-tvos@latest`.

To build from sources, I've added the following to `settings.gradle` as the [docs](https://reactnative.dev/contributing/how-to-build-from-source#update-your-project-to-build-from-source) say (but using this repo's maven `io.github.react-native-tvos`):
```
includeBuild('../node_modules/react-native') {
  dependencySubstitution {
    substitute(module("io.github.react-native-tvos:react-android")).using(project(":packages:react-native:ReactAndroid"))
    substitute(module("io.github.react-native-tvos:react-native")).using(project(":packages:react-native:ReactAndroid"))
    substitute(module("io.github.react-native-tvos:hermes-android")).using(project(":packages:react-native:ReactAndroid:hermes-engine"))
    substitute(module("io.github.react-native-tvos:hermes-engine")).using(project(":packages:react-native:ReactAndroid:hermes-engine"))
  }
}
```

Just try to build it with Android Studio or `yarn android`.

**Comment by douglowder** [comment]
I'll have a look to see what the problem is.

**Comment by douglowder** [comment]
Thanks for finding this. It is happening because I had to do some hacks during the React Native TV build to ensure that Android hermes engine and hermesc was exactly the same as the one used in RN core 0.72.4.

The issue you are having is that your build is picking up hermes from the main branch in the Hermes repo, instead of the correct tag for RN 0.72.4.

I created a repo, starting with your test project: https://github.com/douglowder/test-rn-android-source-build

This patch

https://github.com/douglowder/test-rn-android-source-build/commit/c8300aba940ed622bcaec9cfb0c1f29e98cfc3d5

adds the RN core package as a dev dependency, so you can access the `sdks/.hermesversion` file with the correct Hermes repo tag to use. Then you need a slight change in `ReactAndroid/hermes-engine/build.gradle` to pick that up, and your build will work.

**Comment by leonchabbey** [comment]
Thanks a lot of fixing this so fast! I confirm it works now 👍 

I wonder, could we find a way to automatically detect in what environment the code is running (external project vs react-native-tvos repo) so this patch wouldn't be needed? I have limited experience with Android / gradle so if you have an idea how to do it it'd great otherwise I'll try some stuff when I have the time.

If not, I think we should at least add to the README a "Building from sources" tutorial for people who needs to maintain a fork for their project and want to use react-native 0.72+.

**Comment by freeboub** [comment]
I confirm it is useful to know how to build from sources !

**Comment by douglowder** [comment]
Fixed by adding a bunch of `pickFirst` directives to `android/app/build.gradle`.

https://github.com/douglowder/test-rn-android-source-build/commit/f9615ce493696c869dc29e92110d54eac7f8874a

Might be overkill to add every library in the react-native intermediates directory, but it definitely fixes the build.

**Comment by leonchabbey** [comment]
Yep that fixed the build for development! 👍 

However, there are other issues when trying to build a release bundle/apk.

Those errors are reproductible by using this command:
```
npx react-native build-android --mode=release
---- or ----
cd android && ./gradlew bundleRelease
```

## In https://github.com/douglowder/test-rn-android-source-build

I get errors such as:
```
> Task :app:mergeDexRelease FAILED
ERROR:/Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/sub_project_dex_archive/release/out/c83d3b59cecede5a5a1e1dbab1abe868ea063eb7f97138a82b783b25c3e7ff6a_2.jar: D8: Type com.facebook.debug.debugoverlay.model.DebugOverlayTag is defined multiple times: /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/sub_project_dex_archive/release/out/c83d3b59cecede5a5a1e1dbab1abe868ea063eb7f97138a82b783b25c3e7ff6a_2.jar:classes.dex, /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/external_libs_dex/release/mergeExtDexRelease/classes.dex
com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives:
Learn how to resolve the issue at https://developer.android.com/studio/build/dependencies#duplicate_classes.
Type com.facebook.debug.debugoverlay.model.DebugOverlayTag is defined multiple times: /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/sub_project_dex_archive/release/out/c83d3b59cecede5a5a1e1dbab1abe868ea063eb7f97138a82b783b25c3e7ff6a_2.jar:classes.dex, /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/external_libs_dex/release/mergeExtDexRelease/classes.dex
        at com.android.builder.dexing.D8DexArchiveMerger.getExceptionToRethrow(D8DexArchiveMerger.java:151)
        at com.android.builder.dexing.D8DexArchiveMerger.mergeDexArchives(D8DexArchiveMerger.java:138)
        at com.android.build.gradle.internal.tasks.DexMergingWorkAction.merge(DexMergingTask.kt:858)
        at com.android.build.gradle.internal.tasks.DexMergingWorkAction.run(DexMergingTask.kt:804)
        at com.android.build.gradle.internal.profile.ProfileAwareWorkAction.execute(ProfileAwareWorkAction.kt:74)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.create(NoIsolationWorkerFactory.java:66)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.create(NoIsolationWorkerFactory.java:62)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1.lambda$execute$0(NoIsolationWorkerFactory.java:62)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
        at org.gradle.workers.internal.AbstractWorker.executeWrappedInBuildOperation(AbstractWorker.java:41)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1.execute(NoIsolationWorkerFactory.java:59)
        at org.gradle.workers.internal.DefaultWorkerExecutor.lambda$submitWork$0(DefaultWorkerExecutor.java:169)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:187)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.access$700(DefaultConditionalExecutionQueue.java:120)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner$1.run(DefaultConditionalExecutionQueue.java:162)
        at org.gradle.internal.Factories$1.create(Factories.java:31)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:249)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:109)
        at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:114)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:157)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:126)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.android.tools.r8.CompilationFailedException: Compilation failed to complete, origin: /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/sub_project_dex_archive/release/out/c83d3b59cecede5a5a1e1dbab1abe868ea063eb7f97138a82b783b25c3e7ff6a_2.jar:classes.dex
        at Version.fakeStackEntry(Version_4.0.52.java:0)
        at com.android.tools.r8.internal.vk.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:75)
        at com.android.tools.r8.internal.vk.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:28)
        at com.android.tools.r8.internal.vk.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:27)
        at com.android.tools.r8.internal.vk.b(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:2)
        at com.android.tools.r8.D8.run(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:11)
        at com.android.builder.dexing.D8DexArchiveMerger.mergeDexArchives(D8DexArchiveMerger.java:136)
        ... 38 more
Caused by: com.android.tools.r8.internal.g: Type com.facebook.debug.debugoverlay.model.DebugOverlayTag is defined multiple times: /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/sub_project_dex_archive/release/out/c83d3b59cecede5a5a1e1dbab1abe868ea063eb7f97138a82b783b25c3e7ff6a_2.jar:classes.dex, /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/external_libs_dex/release/mergeExtDexRelease/classes.dex
        at com.android.tools.r8.internal.GV.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:14)
        at com.android.tools.r8.internal.GV.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:22)
        at com.android.tools.r8.internal.rP.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:44)
        at com.android.tools.r8.internal.rP.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:10)
        at java.base/java.util.concurrent.ConcurrentHashMap.merge(ConcurrentHashMap.java:2048)
        at com.android.tools.r8.internal.rP.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:6)
        at com.android.tools.r8.graph.F2$a.e(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:7)
        at com.android.tools.r8.dex.b.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:84)
        at com.android.tools.r8.dex.b.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:23)
        at com.android.tools.r8.dex.b.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:22)
        at com.android.tools.r8.D8.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:50)
        at com.android.tools.r8.D8.d(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:19)
        at com.android.tools.r8.D8.b(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:1)
        at com.android.tools.r8.internal.vk.a(R8_4.0.52_5a340ca2823c7e792fe09805c75f749b9d398d230bc0518bb54ae9b6b50addbe:24)
        ... 41 more
----------------------------------------------------
Caused by: com.android.tools.r8.internal.g: Type com.facebook.debug.debugoverlay.model.DebugOverlayTag is defined multiple times: /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/sub_project_dex_archive/release/out/c83d3b59cecede5a5a1e1dbab1abe868ea063eb7f97138a82b783b25c3e7ff6a_2.jar:classes.dex, /Users/leon/Documents/projects/test-rn-android-source-build/android/app/build/intermediates/external_libs_dex/release/mergeExtDexRelease/classes.dex
------------------------------------
```

Could this be linked to having added `react-native` a second time as `react-native-core`?

## In my own project
Not sure why but the first error above never occured in this project...
Here, I get the following error:
```
Couldn't determine Hermesc location. Please set `react.hermesCommand` to the path of the hermesc binary file. node_modules/react-native/sdks/hermesc/%OS-BIN%/hermesc
```
To my knowledge, I haven't touched anything related to this. My project should have the same setup as your test project. (and is also not a monorepo FYI).
Maybe this error will also occur in the test repo if it wouldn't crash earlier with the first error above?

Thanks for the help debugging all of this!

**Comment by douglowder** [comment]
Sorry for the delay @leonchabbey I'll get to this again soon

**Comment by leonchabbey** [comment]
Hi @douglowder!

With the help of @SebastienChauvin, we found a way to build and bundle when bundling from sources.
It's still very hacky though. The fixes for the issues reported in my last message were:
1. Reference `hermesc` from `react-native-core`
2. Remove references/implementations of react-native remote repository in `android/app/build.gradle` to avoid duplicate stuffs during DEX merging.

Here is the commit with all necessary changes to build from sources: https://github.com/leonchabbey/test-rn-android-source-build/commit/80a7d103fb9990311c28d3c4940dc1c348774d00
Keep in mind what needs to be added/removed/patched depends on what the developer wants to do (debug native code, bundle app, ...)

Hopefully these fixes allow everyone to build from sources until a more clean and permanent solution is found!

**Comment by douglowder** [comment]
@leonchabbey THANK YOU!!!

I have noticed that with 0.73, there is a nice property that you can add to `android/gradle.properties` that will cause your build to use local Maven artifacts in `/tmp/maven-local` instead of the published artifacts. So you could start with RN source, make your changes, publish to the local Maven repo on your machine, and then build your project just as you would in production. I think this would probably be less hacky, but would be slightly inconvenient (react-native build would be separate from your application build).

**Comment by jhen0409** [comment]
Thanks for sharing solution!

Also, I've found I need to use `compileOnly("io.github.react-native-tvos:react-android")` in `android/app/build.gradle` for release build to made it works. (for https://github.com/leonchabbey/test-rn-android-source-build/commit/80a7d103fb9990311c28d3c4940dc1c348774d00, it also works in debug mode)

**Comment by robingullo** [comment]
@douglowder Thank you for the example repo, it helped me a lot! It did not take source file modifications into account, but I made it work by changing the dependencies in `settings.gradle` :
```diff
includeBuild('../node_modules/react-native') {
  dependencySubstitution {
    substitute(module("io.github.react-native-tvos:react-android")).using(project(":packages:react-native:ReactAndroid"))
-     substitute(module("io.github.react-native-tvos:react-native")).using(project(":packages:react-native:ReactAndroid"))
+     substitute(module("com.facebook.react:react-native")).using(project(":packages:react-native:ReactAndroid"))
    substitute(module("io.github.react-native-tvos:hermes-android")).using(project(":packages:react-native:ReactAndroid:hermes-engine"))
-     substitute(module("io.github.react-native-tvos:hermes-engine")).using(project(":packages:react-native:ReactAndroid:hermes-engine"))
+     substitute(module("com.facebook.react:hermes-engine")).using(project(":packages:react-native:ReactAndroid:hermes-engine"))
  }
}
```

**Comment by DrRefactor** [comment]
Hi @douglowder @leonchabbey 

I'm trying to build react-native-tvos 0.72.6-1 with your fixes, it fails with following:
```
> Task :app:processAndroidLauncherDebugMainManifest FAILED
[:packages:react-native:ReactAndroid] /Users/pawelwieckowski/Projects/com/nitrox5/node_modules/react-native/ReactAndroid/build/intermediates/merged_manifest/debug/AndroidManifest.xml Error:
        Namespace 'com.facebook.react' used in: :packages:react-native:ReactAndroid, io.github.react-native-tvos:react-android:0.72.6-1.
[:packages:react-native:ReactAndroid:hermes-engine] /Users/pawelwieckowski/Projects/com/nitrox5/node_modules/react-native/ReactAndroid/hermes-engine/build/intermediates/merged_manifest/debug/AndroidManifest.xml Error:
        Namespace 'com.facebook.hermes' used in: :packages:react-native:ReactAndroid:hermes-engine, io.github.react-native-tvos:hermes-android:0.72.6-1.
/Users/pawelwieckowski/Projects/com/nitrox5/packages/native/android/app/src/debug/AndroidManifest.xml Error:
        Validation failed, exiting
```

Have you seen this error by any chance, or do you know what might be wrong here? I applied all diffs from @leonchabbey repo.

**Comment by douglowder** [comment]
Circling back to this issue, I am now tracking this in #843 , and the issue does seem to be resolved in RNTV 0.76.5-0 and later.

I created a new Expo TV project using RNTV 0.76.5-0, executed prebuild to create the android directory, and then added these lines to the end of `android/settings.gradle`:

```gradle
includeBuild('../node_modules/react-native') {
    dependencySubstitution {
        substitute(module('io.github.react-native-tvos:react-android')).using(project(':packages:react-native:ReactAndroid'))
        substitute(module('com.facebook.react:react-android')).using(project(':packages:react-native:ReactAndroid'))
        substitute(module('com.facebook.react:react-native')).using(project(':packages:react-native:ReactAndroid'))

        substitute(module('io.github.react-native-tvos:hermes-android')).using(project(':packages:react-native:ReactAndroid:hermes-engine'))
        substitute(module('com.facebook.react:hermes-android')).using(project(':packages:react-native:ReactAndroid:hermes-engine'))
        substitute(module('com.facebook.react:hermes-engine')).using(project(':packages:react-native:ReactAndroid:hermes-engine'))
    }
}
```

With this change, build from source works correctly with Android Studio.
